src/substrait/textplan/PlanPrinterVisitor.cpp

```diff
@@ -177,6 +190,65 @@ std::string PlanPrinterVisitor::lookupFieldReference(
   return symbol->name;
 }
 
+std::string PlanPrinterVisitor::lookupFieldReferenceForEmit(
+    uint32_t fieldReference,
+    const SymbolInfo* currentScope,
+    uint32_t stepsOut,
+    bool needFullyQualified) {
+  if (currentScope == nullptr || *currentScope_ == SymbolInfo::kUnknown) {
+    errorListener_->addError(
+        "Field number " + std::to_string(fieldReference) +
+        " mysteriously requested outside of a relation.");
+    return "field#" + std::to_string(fieldReference);
+  }
+  auto actualScope = currentScope;
+  if (stepsOut > 0) {
+    for (auto stepsLeft = stepsOut; stepsLeft > 0; stepsLeft--) {
+      auto actualParentQueryLocation = getParentQueryLocation(actualScope);
+      if (actualParentQueryLocation == Location::kUnknownLocation) {
+        errorListener_->addError(
+            "Requested steps out of " + std::to_string(stepsOut) +
+            " but not within subquery depth that high.");
+        return "field#" + std::to_string(fieldReference);
+      }
+      actualScope = symbolTable_->lookupSymbolByLocationAndType(
+          actualParentQueryLocation, SymbolType::kRelation);
+      if (actualScope == nullptr) {
+        errorListener_->addError(
+            "Internal error: Missing previously encountered parent query symbol.");
+        return "field#" + std::to_string(fieldReference);
+      }
+    }
+  }
+  auto relationData =
+      ANY_CAST(std::shared_ptr<RelationData>, actualScope->blob);
+  const SymbolInfo* symbol{nullptr};
+  auto fieldReferencesSize = relationData->fieldReferences.size();
+  if (isAggregate(currentScope) &&
+      fieldReference < relationData->generatedFieldReferences.size()) {
+    symbol = relationData->generatedFieldReferences[fieldReference];
+  } else if (fieldReference < fieldReferencesSize) {
+    symbol = relationData->fieldReferences[fieldReference];
+  } else if (
+      fieldReference <
+      fieldReferencesSize + relationData->generatedFieldReferences.size()) {
+    symbol =
+        relationData
+            ->generatedFieldReferences[fieldReference - fieldReferencesSize];
+  } else {
+    errorListener_->addError(
+        "Encountered field reference out of range: " +
+        std::to_string(fieldReference));
+    return "field#" + std::to_string(fieldReference);
+  }
```

Comment:
This seems odd to me because I'd expect a clean outer branch on the type of `currentScope` *before* we look at the value of fieldReference. If currentScope *is* an aggregate but fieldReference is out of range, is it really reasonable to use the same code we would have used if currentScope were not an aggregate? I'd expect something like:
```suggestion
  if (isAggregate(currentScope)) {
    if (fieldReference < relationData->generatedFieldReferences.size()) {
      symbol = relationData->generatedFieldReferences[fieldReference];
    }
  } else {
    auto size = relationData->fieldReferences.size();
    if (fieldReference < size) {
      symbol = relationData->fieldReferences[fieldReference];
    } else if (
        fieldReference <
        size + relationData->generatedFieldReferences.size()) {
      symbol =
          relationData
              ->generatedFieldReferences[fieldReference - size];
    }
  }
  if (symbol == nullptr) {
    errorListener_->addError(
        "Encountered field reference out of range: " +
        std::to_string(fieldReference));
    return "field#" + std::to_string(fieldReference);
  }
```

in any case, please clarify the branching here

Reply:
Agreed, much more understandable with your approach.